### PR TITLE
Performance improvement for some libcudf regex functions for long strings

### PR DESCRIPTION
**Pull request by davidwendt**
## Description
Changes the internal regex logic to minimize character counting to help performance with longer strings. The improvement applies mainly to libcudf regex functions that return strings (i.e. extract, replace, split). The changes here also improve the internal device APIs for clarity to improve maintenance. The most significant change makes the position variables input-only and returning an optional pair to indicate a successful match.

There are some more optimizations that are possible here where character positions are passed back and forth that could be replaced with byte positions to further reduce counting. Initial measurements showed this noticeably slowed down small strings so more analysis is required before continuing this optimization. 

Reference: https://github.com/rapidsai/cudf/pull/13480

### More Detail

First, there is a change to some internal regex function signatures. Notable the `reprog_device::find()` and `reprog_device::extract()` member functions declared in `cpp/src/strings/regex/regex.cuh` that are used by all the libcudf regex functions. The in/out parameters are now input-only parameters (pass by value) and the return is an optional pair that includes the match result. Also, the `begin` parameter is now an iterator and the `end` parameter now has a default. This change requires updating all the definitions and uses of the `find` and `extract` member functions.

Using an iterator as the `begin` parameter allows for some optimizations in the calling code to minimize character counting that may be needed for processing multi-byte UTF-8 characters. Rather than using the `cudf::string_view::byte_offset()` member function to convert character positions to byte positions, an iterator can be incremented as we traverse through the string which helps reduce some character counting. So the changes here involve removing some calls to `byte_offset()` and incrementing (really moving) iterators with a pattern like `itr += (new_pos - itr.position());` There is another PR #13428 to make a `move_to` iterator member function.

It is possible to reduce the character counting even more as mentioned above but further optimization requires some deeper analysis.

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

**Comment by davidwendt** [comment]
Partial benchmark results for `extract`:
```
| width |    rows  |   baseline |   new time |          Diff |   %Diff |
|-------|----------|------------|------------|---------------|---------|
|   32  |   4096   | 102.248 us | 101.055 us |     -1.193 us |  -1.17% |
|   64  |   4096   | 104.781 us | 101.499 us |     -3.282 us |  -3.13% |
|   128 |   4096   | 120.856 us | 109.923 us |    -10.934 us |  -9.05% |
|   256 |   4096   | 151.494 us | 118.979 us |    -32.515 us | -21.46% |
|   512 |   4096   | 239.063 us | 139.072 us |    -99.992 us | -41.83% |
|  1024 |   4096   | 393.707 us | 178.753 us |   -214.954 us | -54.60% |
|  2048 |   4096   |   1.161 ms | 250.995 us |   -910.083 us | -78.38% |
|   32  |  32768   | 119.949 us | 116.893 us |     -3.056 us |  -2.55% |
|   64  |  32768   | 131.676 us | 121.664 us |    -10.012 us |  -7.60% |
|   128 |  32768   | 159.921 us | 135.877 us |    -24.044 us | -15.03% |
|   256 |  32768   | 272.049 us | 160.928 us |   -111.120 us | -40.85% |
|   512 |  32768   | 472.972 us | 200.578 us |   -272.395 us | -57.59% |
|  1024 |  32768   | 817.860 us | 279.889 us |   -537.971 us | -65.78% |
|  2048 |  32768   |   1.513 ms | 435.822 us |  -1077.136 us | -71.19% |
|   32  |  262144  | 406.995 us | 391.527 us |    -15.468 us |  -3.80% |
|   64  |  262144  | 471.935 us | 425.180 us |    -46.755 us |  -9.91% |
|   128 |  262144  | 706.503 us | 543.007 us |   -163.496 us | -23.14% |
|   256 |  262144  |   1.471 ms | 727.336 us |   -743.367 us | -50.55% |
|   512 |  262144  |   2.592 ms | 986.388 us |  -1605.897 us | -61.95% |
|  1024 |  262144  |   5.092 ms |   1.662 ms |  -3429.774 us | -67.36% |
|  2048 |  262144  |  10.022 ms |   2.850 ms |  -7172.154 us | -71.56% |
|   32  | 2097152  |   2.689 ms |   2.569 ms |   -120.485 us |  -4.48% |
|   64  | 2097152  |   3.153 ms |   2.845 ms |   -308.122 us |  -9.77% |
|   128 | 2097152  |   4.857 ms |   3.707 ms |  -1149.617 us | -23.67% |
|   256 | 2097152  |  10.554 ms |   5.060 ms |  -5494.176 us | -52.06% |
|   512 | 2097152  |  19.063 ms |   6.936 ms | -12127.088 us | -63.62% |
|   32  | 16777216 |  21.228 ms |  20.336 ms |   -892.067 us |  -4.20% |
|   64  | 16777216 |  24.840 ms |  22.556 ms |  -2283.420 us |  -9.19% |
```
(it's a bit frustrating that ~3x improvement shows up as ~70% diff)

**Comment by davidwendt** [comment]
Benchmark results for `replace_re`
```
|  width |    rows  |   Ref Time |   New Time |           Diff |   %Diff |
|--------|----------|------------|------------|----------------|---------|
|   32   |  32768   | 486.621 us | 495.978 us |       9.357 us |   1.92% |
|   64   |  32768   |   1.017 ms | 903.366 us |    -113.848 us | -11.19% |
|   128  |  32768   |   2.153 ms |   1.925 ms |    -228.080 us | -10.59% |
|   256  |  32768   |   4.830 ms |   3.669 ms |   -1161.246 us | -24.04% |
|   512  |  32768   |  13.340 ms |   8.478 ms |   -4862.047 us | -36.45% |
|   32   |  262144  |   2.337 ms |   2.226 ms |    -111.286 us |  -4.76% |
|   64   |  262144  |   5.012 ms |   4.522 ms |    -489.288 us |  -9.76% |
|   128  |  262144  |  11.232 ms |   9.459 ms |   -1772.463 us | -15.78% |
|   256  |  262144  |  26.323 ms |  19.181 ms |   -7141.681 us | -27.13% |
|   512  |  262144  |  73.040 ms |  39.630 ms |  -33409.659 us | -45.74% |
|   32   | 2097152  |  16.935 ms |  15.828 ms |   -1106.769 us |  -6.54% |
|   64   | 2097152  |  36.327 ms |  32.613 ms |   -3714.002 us | -10.22% |
|   128  | 2097152  |  81.700 ms |  68.062 ms |  -13638.190 us | -16.69% |
|   256  | 2097152  | 193.886 ms | 140.669 ms |  -53217.281 us | -27.45% |
|   512  | 2097152  | 532.759 ms | 288.650 ms | -244109.308 us | -45.82% |
|   32   | 16777216 | 135.575 ms | 126.735 ms |   -8840.007 us |  -6.52% |
|   64   | 16777216 | 290.647 ms | 260.998 ms |  -29648.809 us | -10.20% |
```

And for `replace_with_backrefs`:
```
|  width |    rows  |   Ref Time |   New Time |           Diff |   %Diff |
|--------|----------|------------|------------|----------------|---------|
|   32   |  32768   | 681.210 us | 645.954 us |     -35.256 us |  -5.18% |
|   64   |  32768   |   1.329 ms |   1.118 ms |    -211.121 us | -15.88% |
|   128  |  32768   |   2.896 ms |   2.207 ms |    -688.475 us | -23.78% |
|   256  |  32768   |   7.039 ms |   4.148 ms |   -2890.664 us | -41.07% |
|   512  |  32768   |  21.115 ms |   9.486 ms |  -11629.112 us | -55.07% |
|   32   |  262144  |   3.019 ms |   2.825 ms |    -194.372 us |  -6.44% |
|   64   |  262144  |   6.411 ms |   5.557 ms |    -854.095 us | -13.32% |
|   128  |  262144  |  14.713 ms |  11.233 ms |   -3479.592 us | -23.65% |
|   256  |  262144  |  36.674 ms |  22.629 ms |  -14045.267 us | -38.30% |
|   512  |  262144  | 112.022 ms |  46.204 ms |  -65818.233 us | -58.75% |
|   32   | 2097152  |  21.612 ms |  20.283 ms |   -1329.841 us |  -6.15% |
|   64   | 2097152  |  46.670 ms |  40.522 ms |   -6148.734 us | -13.17% |
|   128  | 2097152  | 107.527 ms |  82.741 ms |  -24786.012 us | -23.05% |
|   256  | 2097152  | 271.823 ms | 168.984 ms | -102838.835 us | -37.83% |
|   512  | 2097152  | 817.977 ms | 343.122 ms | -474855.330 us | -58.05% |
|   32   | 16777216 | 172.788 ms | 162.064 ms |  -10724.165 us |  -6.21% |
|   64   | 16777216 | 372.764 ms | 323.421 ms |  -49343.678 us | -13.24% |
```

**Comment by davidwendt** [comment]
Benchmark results for `split_re`:
```
| width |    rows  |   Ref Time |   New Time |          Diff |   %Diff |
|-------|----------|------------|------------|---------------|---------|
|   32  |   4096   | 140.207 us | 139.857 us |     -0.350 us |  -0.25% |
|   64  |   4096   | 299.160 us | 293.625 us |     -5.536 us |  -1.85% |
|   128 |   4096   | 545.842 us | 521.533 us |    -24.309 us |  -4.45% |
|   256 |   4096   |   1.193 ms |   1.042 ms |   -151.348 us | -12.69% |
|   512 |   4096   |   2.601 ms |   2.009 ms |   -591.602 us | -22.75% |
|  1024 |   4096   |   6.674 ms |   4.177 ms |  -2497.314 us | -37.42% |
|  2048 |   4096   |  19.341 ms |   8.332 ms | -11009.056 us | -56.92% |
|   32  |  32768   | 163.030 us | 160.657 us |     -2.373 us |  -1.46% |
|   64  |  32768   | 348.338 us | 354.877 us |      6.539 us |   1.88% |
|   128 |  32768   | 706.315 us | 670.118 us |    -36.197 us |  -5.12% |
|   256 |  32768   |   1.401 ms |   1.221 ms |   -180.521 us | -12.88% |
|   512 |  32768   |   3.382 ms |   2.614 ms |   -768.344 us | -22.72% |
|  1024 |  32768   |   9.380 ms |   5.201 ms |  -4178.640 us | -44.55% |
|  2048 |  32768   |  29.272 ms |   9.841 ms | -19431.435 us | -66.38% |
|   32  |  262144  | 656.388 us | 663.743 us |      7.355 us |   1.12% |
|   64  |  262144  |   1.417 ms |   1.398 ms |    -18.677 us |  -1.32% |
|   128 |  262144  |   3.034 ms |   2.879 ms |   -154.749 us |  -5.10% |
|   256 |  262144  |   7.015 ms |   6.028 ms |   -987.029 us | -14.07% |
|   512 |  262144  |  17.794 ms |  12.679 ms |  -5114.886 us | -28.74% |
|  1024 |  262144  |  48.859 ms |  25.731 ms | -23128.494 us | -47.34% |
|  2048 |  262144  | 148.964 ms |  51.975 ms | -96989.264 us | -65.11% |
|   32  | 2097152  |   4.788 ms |   4.900 ms |    112.193 us |   2.34% |
|   64  | 2097152  |  10.185 ms |  10.209 ms |     23.962 us |   0.24% |
|   128 | 2097152  |  22.024 ms |  21.336 ms |   -688.066 us |  -3.12% |
|   256 | 2097152  |  51.392 ms |  44.852 ms |  -6540.513 us | -12.73% |
|   512 | 2097152  | 129.414 ms |  93.043 ms | -36371.177 us | -28.10% |
|   32  | 16777216 |  37.788 ms |  38.687 ms |    898.768 us |   2.38% |
|   64  | 16777216 |  80.563 ms |  81.020 ms |    456.425 us |   0.57% |

```

**Review by harrism — COMMENTED** [review]
There are a lot of disparate changes here. I can't review them in detail, but I wonder what is the relationship between them all? Is there something common between the optimizations?

**Comment by davidwendt** [comment]
> There are a lot of disparate changes here. I can't review them in detail, but I wonder what is the relationship between them all? Is there something common between the optimizations?

Yes. There are two main changes here. One is the change to the signatures of the two `reprog_device` member functions `find()` and `extract()` and this allows for the 2nd main change which is using the `string_view::const_iterator` more instead of the `string_view::byte_offset()` to help minimize character counting.  I've also added more detail about both of these changes in the PR description under the `More Detail` heading.

**Review by harrism — APPROVED** [review]
Nice speedups!

**Comment by mythrocks** [comment]
I'm reviewing this right now.

**Review by mythrocks — COMMENTED** [review]
A couple of clarifications and nitpicks. Looks good to me.

File: cpp/include/cudf/strings/detail/utilities.cuh [hunk 1]
```diff
@@ -29,14 +32,15 @@ namespace detail {
...
+ * @param buffer Device buffer to copy to
+ * @param input Data to copy from
+ * @param bytes Number of bytes to copy
+ * @return Pointer to the end of the output buffer after the copy
  */
 __device__ inline char* copy_and_increment(char* buffer, char const* input, size_type bytes)
 {
-  memcpy(buffer, input, bytes);
+  // this can be slightly faster than memcpy
+  thrust::copy_n(thrust::seq, input, bytes, buffer);
```
[the remark this answers is not in the document]
Comment (davidwendt) [reply cpp/include/cudf/strings/detail/utilities.cuh:43]: I'm exactly sure. It seems to _sometimes_ help the GPU utilize the L1/L2 cache better according to the ncu profiling.

File: cpp/src/strings/contains.cu [hunk 2]
```diff
@@ -50,10 +50,9 @@ struct contains_fn {
     if (d_strings.is_null(idx)) return false;
     auto const d_str = d_strings.element<string_view>(idx);
 
-    size_type begin = 0;
-    size_type end   = beginning_only ? 1    // match only the beginning of the string;
-                                     : -1;  // match anywhere in the string
-    return static_cast<bool>(prog.find(thread_idx, d_str, begin, end));
+    size_type end = beginning_only ? 1    // match only the beginning of the string;
+                                   : -1;  // match anywhere in the string
+    return prog.find(thread_idx, d_str, d_str.begin(), end).has_value();
```
[the remark this answers is not in the document]
Comment (davidwendt) [reply cpp/src/strings/contains.cu:55]: No, that produces a compiler error.

File: cpp/src/strings/regex/regex.cuh [hunk 3]
```diff
@@ -184,36 +182,33 @@ class reprog_device {
...
    * @param thread_idx The index used for mapping the state memory for this string in global memory.
    * @param d_str The string to search.
-   * @param[in,out] begin Position index to begin the search. If found, returns the position found
-   * in the string.
-   * @param[in,out] end Position index to end the search. If found, returns the last position
-   * matching in the string.
-   * @return Returns 0 if no match is found.
+   * @param begin Position to begin the search within `d_str`.
+   * @param end Character position index to end the search within `d_str`.
+   *            Specify -1 to match any virtual positions past the end of the string.
```
[the remark this answers is not in the document]
Comment (davidwendt) [reply cpp/src/strings/regex/regex.cuh:187]: The end of the string is technically one position past the last character of the string.

**Review by mythrocks — APPROVED** [review]

File: cpp/include/cudf/strings/detail/utilities.cuh (hunk 1 above)
[the remark this answers is not in the document]
Comment (harrism) [reply cpp/include/cudf/strings/detail/utilities.cuh:43]: Hopefully thrust::copy_n knows when it should just used memcpy. :)  Performance might be GPU-dependent.

**Pull request merged**